// typescript-rest/dist/server-container.js
'use strict';
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var bodyParser = require("body-parser");
var cookieParser = require("cookie-parser");
var _ = require("lodash");
var multer = require("multer");
var metadata = require("./metadata");
var Errors = require("./server-errors");
var server_return_1 = require("./server-return");
var server_types_1 = require("./server-types");
var InternalServer = /** @class */ (function () {
    function InternalServer(router) {
        this.router = router;
    }
    InternalServer.registerServiceClass = function (target) {
        InternalServer.pathsResolved = false;
        target = InternalServer.serviceFactory.getTargetClass(target);
        var name = target['name'] || target.constructor['name'];
        if (!InternalServer.serverClasses.has(name)) {
            InternalServer.serverClasses.set(name, new metadata.ServiceClass(target));
            InternalServer.inheritParentClass(name);
        }
        var serviceClass = InternalServer.serverClasses.get(name);
        return serviceClass;
    };
    InternalServer.inheritParentClass = function (name) {
        var classData = InternalServer.serverClasses.get(name);
        var parent = Object.getPrototypeOf(classData.targetClass.prototype).constructor;
        var parentClassData = InternalServer.getServiceClass(parent);
        if (parentClassData) {
            if (parentClassData.methods) {
                parentClassData.methods.forEach(function (value, key) {
                    classData.methods.set(key, _.cloneDeep(value));
                });
            }
            if (parentClassData.properties) {
                parentClassData.properties.forEach(function (value, key) {
                    classData.properties.set(key, _.cloneDeep(value));
                });
            }
            if (parentClassData.languages) {
                for (var _i = 0, _a = parentClassData.languages; _i < _a.length; _i++) {
                    var lang = _a[_i];
                    classData.languages.push(lang);
                }
            }
            if (parentClassData.accepts) {
                for (var _b = 0, _c = parentClassData.accepts; _b < _c.length; _b++) {
                    var acc = _c[_b];
                    classData.accepts.push(acc);
                }
            }
        }
    };
    InternalServer.registerServiceMethod = function (target, methodName) {
        if (methodName) {
            InternalServer.pathsResolved = false;
            var classData = InternalServer.registerServiceClass(target);
            if (!classData.methods.has(methodName)) {
                classData.methods.set(methodName, new metadata.ServiceMethod());
            }
            var serviceMethod = classData.methods.get(methodName);
            return serviceMethod;
        }
        return null;
    };
    InternalServer.resolveAllPaths = function () {
        if (!InternalServer.pathsResolved) {
            InternalServer.paths.clear();
            InternalServer.serverClasses.forEach(function (classData) {
                classData.methods.forEach(function (method) {
                    if (!method.resolvedPath) {
                        InternalServer.resolveProperties(classData, method);
                    }
                });
            });
            InternalServer.pathsResolved = true;
        }
    };
    InternalServer.getPaths = function () {
        InternalServer.resolveAllPaths();
        var result = new Set();
        InternalServer.paths.forEach(function (value, key) {
            result.add(key);
        });
        return result;
    };
    InternalServer.getHttpMethods = function (path) {
        InternalServer.resolveAllPaths();
        var methods = InternalServer.paths.get(path);
        return methods || new Set();
    };
    InternalServer.getServiceClass = function (target) {
        target = InternalServer.serviceFactory.getTargetClass(target);
        return InternalServer.serverClasses.get(target['name'] || target.constructor['name']) || null;
    };
    InternalServer.resolveLanguages = function (serviceClass, serviceMethod) {
        var resolvedLanguages = new Array();
        if (serviceClass.languages) {
            serviceClass.languages.forEach(function (lang) {
                resolvedLanguages.push(lang);
            });
        }
        if (serviceMethod.languages) {
            serviceMethod.languages.forEach(function (lang) {
                resolvedLanguages.push(lang);
            });
        }
        if (resolvedLanguages.length > 0) {
            serviceMethod.resolvedLanguages = resolvedLanguages;
        }
    };
    InternalServer.resolveAccepts = function (serviceClass, serviceMethod) {
        var resolvedAccepts = new Array();
        if (serviceClass.accepts) {
            serviceClass.accepts.forEach(function (accept) {
                resolvedAccepts.push(accept);
            });
        }
        if (serviceMethod.accepts) {
            serviceMethod.accepts.forEach(function (accept) {
                resolvedAccepts.push(accept);
            });
        }
        if (resolvedAccepts.length > 0) {
            serviceMethod.resolvedAccepts = resolvedAccepts;
        }
    };
    InternalServer.resolveProperties = function (serviceClass, serviceMethod) {
        InternalServer.resolveLanguages(serviceClass, serviceMethod);
        InternalServer.resolveAccepts(serviceClass, serviceMethod);
        InternalServer.resolvePath(serviceClass, serviceMethod);
    };
    InternalServer.resolvePath = function (serviceClass, serviceMethod) {
        var classPath = serviceClass.path ? serviceClass.path.trim() : '';
        var resolvedPath = _.startsWith(classPath, '/') ? classPath : '/' + classPath;
        if (_.endsWith(resolvedPath, '/')) {
            resolvedPath = resolvedPath.slice(0, resolvedPath.length - 1);
        }
        if (serviceMethod.path) {
            var methodPath = serviceMethod.path.trim();
            resolvedPath = resolvedPath + (_.startsWith(methodPath, '/') ? methodPath : '/' + methodPath);
        }
        var declaredHttpMethods = InternalServer.paths.get(resolvedPath);
        if (!declaredHttpMethods) {
            declaredHttpMethods = new Set();
            InternalServer.paths.set(resolvedPath, declaredHttpMethods);
        }
        if (declaredHttpMethods.has(serviceMethod.httpMethod)) {
            throw Error("Duplicated declaration for path [" + resolvedPath + "], method [" + serviceMethod.httpMethod + "].");
        }
        declaredHttpMethods.add(serviceMethod.httpMethod);
        serviceMethod.resolvedPath = resolvedPath;
    };
    InternalServer.prototype.buildServices = function (types) {
        var _this = this;
        if (types) {
            types = types.map(function (type) { return InternalServer.serviceFactory.getTargetClass(type); });
        }
        if (InternalServer.authenticator) {
            InternalServer.authenticator.initialize(this.router);
        }
        InternalServer.serverClasses.forEach(function (classData) {
            if (!classData.isAbstract) {
                classData.methods.forEach(function (method) {
                    if (_this.validateTargetType(classData.targetClass, types)) {
                        _this.buildService(classData, method);
                    }
                });
            }
        });
        InternalServer.pathsResolved = true;
        this.handleNotAllowedMethods();
    };
    InternalServer.prototype.runPreprocessors = function (processors, req) {
        return __awaiter(this, void 0, void 0, function () {
            var _i, processors_1, processor;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _i = 0, processors_1 = processors;
                        _a.label = 1;
                    case 1:
                        if (!(_i < processors_1.length)) return [3 /*break*/, 4];
                        processor = processors_1[_i];
                        return [4 /*yield*/, Promise.resolve(processor(req))];
                    case 2:
                        _a.sent();
                        _a.label = 3;
                    case 3:
                        _i++;
                        return [3 /*break*/, 1];
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    InternalServer.prototype.buildService = function (serviceClass, serviceMethod) {
        if (!serviceMethod.resolvedPath) {
            InternalServer.resolveProperties(serviceClass, serviceMethod);
        }
        var args = [serviceMethod.resolvedPath];
        args = args.concat(this.buildSecurityMiddlewares(serviceClass, serviceMethod));
        args = args.concat(this.buildParserMiddlewares(serviceMethod));
        args.push(this.buildServiceMiddleware(serviceMethod, serviceClass));
        switch (serviceMethod.httpMethod) {
            case server_types_1.HttpMethod.GET:
                this.router.get.apply(this.router, args);
                break;
            case server_types_1.HttpMethod.POST:
                this.router.post.apply(this.router, args);
                break;
            case server_types_1.HttpMethod.PUT:
                this.router.put.apply(this.router, args);
                break;
            case server_types_1.HttpMethod.DELETE:
                this.router.delete.apply(this.router, args);
                break;
            case server_types_1.HttpMethod.HEAD:
                this.router.head.apply(this.router, args);
                break;
            case server_types_1.HttpMethod.OPTIONS:
                this.router.options.apply(this.router, args);
                break;
            case server_types_1.HttpMethod.PATCH:
                this.router.patch.apply(this.router, args);
                break;
            default:
                throw Error("Invalid http method for service [" + serviceMethod.resolvedPath + "]");
        }
    };
    InternalServer.prototype.validateTargetType = function (targetClass, types) {
        if (types && types.length > 0) {
            return (types.indexOf(targetClass) > -1);
        }
        return true;
    };
    InternalServer.prototype.handleNotAllowedMethods = function () {
        var _this = this;
        var paths = InternalServer.getPaths();
        paths.forEach(function (path) {
            var supported = InternalServer.getHttpMethods(path);
            var allowedMethods = new Array();
            supported.forEach(function (method) {
                allowedMethods.push(server_types_1.HttpMethod[method]);
            });
            var allowed = allowedMethods.join(', ');
            _this.router.all(path, function (req, res, next) {
                if (res.headersSent || allowedMethods.indexOf(req.method) > -1) {
                    next();
                }
                else {
                    res.set('Allow', allowed);
                    throw new Errors.MethodNotAllowedError();
                }
            });
        });
    };
    InternalServer.prototype.getUploader = function () {
        if (!this.upload) {
            var options = {};
            if (InternalServer.fileDest) {
                options.dest = InternalServer.fileDest;
            }
            if (InternalServer.fileFilter) {
                options.fileFilter = InternalServer.fileFilter;
            }
            if (InternalServer.fileLimits) {
                options.limits = InternalServer.fileLimits;
            }
            if (options.dest) {
                this.upload = multer(options);
            }
            else {
                this.upload = multer();
            }
        }
        return this.upload;
    };
    InternalServer.prototype.buildServiceMiddleware = function (serviceMethod, serviceClass) {
        var _this = this;
        return function (req, res, next) { return __awaiter(_this, void 0, void 0, function () {
            var allPreprocessors, err_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 4, , 5]);
                        if (!(serviceMethod.processors || serviceClass.processors)) return [3 /*break*/, 2];
                        allPreprocessors = (serviceClass.processors || []).concat(serviceMethod.processors || []);
                        return [4 /*yield*/, this.runPreprocessors(allPreprocessors, req)];
                    case 1:
                        _a.sent();
                        _a.label = 2;
                    case 2: return [4 /*yield*/, this.callTargetEndPoint(serviceClass, serviceMethod, req, res, next)];
                    case 3:
                        _a.sent();
                        next();
                        return [3 /*break*/, 5];
                    case 4:
                        err_1 = _a.sent();
                        next(err_1);
                        return [3 /*break*/, 5];
                    case 5: return [2 /*return*/];
                }
            });
        }); };
    };
    InternalServer.prototype.buildSecurityMiddlewares = function (serviceClass, serviceMethod) {
        var result = new Array();
        var roles = (serviceMethod.roles || []).concat((serviceClass.roles || [])).filter(function (role) { return !!role; });
        if (InternalServer.authenticator && roles.length) {
            result.push(InternalServer.authenticator.getMiddleware());
            roles = roles.filter(function (role) { return role !== '*'; });
            if (roles.length) {
                result.push(function (req, res, next) {
                    var requestRoles = InternalServer.authenticator.getRoles(req);
                    if (requestRoles.some(function (role) { return roles.indexOf(role) >= 0; })) {
                        next();
                    }
                    else {
                        throw new Errors.ForbiddenError();
                    }
                });
            }
        }
        return result;
    };
    InternalServer.prototype.buildParserMiddlewares = function (serviceMethod) {
        var result = new Array();
        if (serviceMethod.mustParseCookies) {
            var args = [];
            if (InternalServer.cookiesSecret) {
                args.push(InternalServer.cookiesSecret);
            }
            if (InternalServer.cookiesDecoder) {
                args.push({ decode: InternalServer.cookiesDecoder });
            }
            result.push(cookieParser.apply(this, args));
        }
        if (serviceMethod.mustParseBody) {
            if (serviceMethod.bodyParserOptions) {
                result.push(bodyParser.json(serviceMethod.bodyParserOptions));
            }
            else {
                result.push(bodyParser.json());
            }
            // TODO adicionar parser de XML para o body
        }
        if (serviceMethod.mustParseForms || serviceMethod.acceptMultiTypedParam) {
            if (serviceMethod.bodyParserOptions) {
                result.push(bodyParser.urlencoded(serviceMethod.bodyParserOptions));
            }
            else {
                result.push(bodyParser.urlencoded({ extended: true }));
            }
        }
        if (serviceMethod.files.length > 0) {
            var options_1 = new Array();
            serviceMethod.files.forEach(function (fileData) {
                if (fileData.singleFile) {
                    options_1.push({ 'name': fileData.name, 'maxCount': 1 });
                }
                else {
                    options_1.push({ 'name': fileData.name });
                }
            });
            result.push(this.getUploader().fields(options_1));
        }
        return result;
    };
    InternalServer.prototype.processResponseHeaders = function (serviceMethod, context) {
        if (serviceMethod.resolvedLanguages) {
            if (serviceMethod.httpMethod === server_types_1.HttpMethod.GET) {
                context.response.vary('Accept-Language');
            }
            context.response.set('Content-Language', context.language);
        }
        if (serviceMethod.resolvedAccepts) {
            context.response.vary('Accept');
        }
    };
    InternalServer.prototype.checkAcceptance = function (serviceMethod, context) {
        if (serviceMethod.resolvedLanguages) {
            var lang = context.request.acceptsLanguages(serviceMethod.resolvedLanguages);
            if (lang) {
                context.language = lang;
            }
        }
        else {
            var languages = context.request.acceptsLanguages();
            if (languages && languages.length > 0) {
                context.language = languages[0];
            }
        }
        if (serviceMethod.resolvedAccepts) {
            var accept = context.request.accepts(serviceMethod.resolvedAccepts);
            if (accept) {
                context.accept = accept;
            }
            else {
                throw new Errors.NotAcceptableError('Accept');
            }
        }
        if (!context.language) {
            throw new Errors.NotAcceptableError('Accept-Language');
        }
    };
    InternalServer.prototype.createService = function (serviceClass, context) {
        var _this = this;
        var serviceObject = InternalServer.serviceFactory.create(serviceClass.targetClass, context);
        if (serviceClass.hasProperties()) {
            serviceClass.properties.forEach(function (property, key) {
                serviceObject[key] = _this.processParameter(property.type, context, property.name, property.propertyType);
            });
        }
        return serviceObject;
    };
    InternalServer.prototype.callTargetEndPoint = function (serviceClass, serviceMethod, req, res, next) {
        return __awaiter(this, void 0, void 0, function () {
            var context, serviceObject, args, toCall, result;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        context = new server_types_1.ServiceContext();
                        context.request = req;
                        context.response = res;
                        context.next = next;
                        this.checkAcceptance(serviceMethod, context);
                        serviceObject = this.createService(serviceClass, context);
                        args = this.buildArgumentsList(serviceMethod, context);
                        toCall = serviceClass.targetClass.prototype[serviceMethod.name] || serviceClass.targetClass[serviceMethod.name];
                        result = toCall.apply(serviceObject, args);
                        this.processResponseHeaders(serviceMethod, context);
                        return [4 /*yield*/, this.sendValue(result, res, next)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    InternalServer.prototype.sendValue = function (value, res, next) {
        return __awaiter(this, void 0, void 0, function () {
            var _a, val;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _a = typeof value;
                        switch (_a) {
                            case 'number': return [3 /*break*/, 1];
                            case 'string': return [3 /*break*/, 2];
                            case 'boolean': return [3 /*break*/, 3];
                            case 'undefined': return [3 /*break*/, 4];
                        }
                        return [3 /*break*/, 5];
                    case 1:
                        res.send(value.toString());
                        return [3 /*break*/, 16];
                    case 2:
                        res.send(value);
                        return [3 /*break*/, 16];
                    case 3:
                        res.send(value.toString());
                        return [3 /*break*/, 16];
                    case 4:
                        if (!res.headersSent) {
                            res.sendStatus(204);
                        }
                        return [3 /*break*/, 16];
                    case 5:
                        if (!(value.filePath && value instanceof server_return_1.DownloadResource)) return [3 /*break*/, 7];
                        return [4 /*yield*/, this.downloadResToPromise(res, value)];
                    case 6:
                        _b.sent();
                        return [3 /*break*/, 16];
                    case 7:
                        if (!(value instanceof server_return_1.DownloadBinaryData)) return [3 /*break*/, 8];
                        if (value.fileName) {
                            res.writeHead(200, {
                                'Content-Length': value.content.length,
                                'Content-Type': value.mimeType,
                                'Content-disposition': 'attachment;filename=' + value.fileName
                            });
                        }
                        else {
                            res.writeHead(200, {
                                'Content-Length': value.content.length,
                                'Content-Type': value.mimeType
                            });
                        }
                        res.end(value.content);
                        return [3 /*break*/, 16];
                    case 8:
                        if (!(value.location && value instanceof server_types_1.ReferencedResource)) return [3 /*break*/, 12];
                        res.set('Location', value.location);
                        if (!value.body) return [3 /*break*/, 10];
                        res.status(value.statusCode);
                        return [4 /*yield*/, this.sendValue(value.body, res, next)];
                    case 9:
                        _b.sent();
                        return [3 /*break*/, 11];
                    case 10:
                        res.sendStatus(value.statusCode);
                        _b.label = 11;
                    case 11: return [3 /*break*/, 16];
                    case 12:
                        if (!(value.then && value.catch)) return [3 /*break*/, 15];
                        return [4 /*yield*/, value];
                    case 13:
                        val = _b.sent();
                        return [4 /*yield*/, this.sendValue(val, res, next)];
                    case 14:
                        _b.sent();
                        return [3 /*break*/, 16];
                    case 15:
                        res.json(value);
                        _b.label = 16;
                    case 16: return [2 /*return*/];
                }
            });
        });
    };
    InternalServer.prototype.downloadResToPromise = function (res, value) {
        return new Promise(function (resolve, reject) {
            res.download(value.filePath, value.filePath, function (err) {
                if (err) {
                    reject(err);
                }
                else {
                    resolve();
                }
            });
        });
    };
    InternalServer.prototype.buildArgumentsList = function (serviceMethod, context) {
        var _this = this;
        var result = new Array();
        serviceMethod.parameters.forEach(function (param) {
            result.push(_this.processParameter(param.paramType, context, param.name, param.type));
        });
        return result;
    };
    InternalServer.prototype.processParameter = function (paramType, context, name, type) {
        switch (paramType) {
            case metadata.ParamType.path:
                return this.convertType(context.request.params[name], type);
            case metadata.ParamType.query:
                return this.convertType(context.request.query[name], type);
            case metadata.ParamType.header:
                return this.convertType(context.request.header(name), type);
            case metadata.ParamType.cookie:
                return this.convertType(context.request.cookies[name], type);
            case metadata.ParamType.body:
                return this.convertType(context.request.body, type);
            case metadata.ParamType.file:
                // @ts-ignore
                var files = context.request.files ? context.request.files[name] : null;
                if (files && files.length > 0) {
                    return files[0];
                }
                return null;
            case metadata.ParamType.files:
                // @ts-ignore
                return context.request.files[name];
            case metadata.ParamType.form:
                return this.convertType(context.request.body[name], type);
            case metadata.ParamType.param:
                var paramValue = context.request.body[name] ||
                    context.request.query[name];
                return this.convertType(paramValue, type);
            case metadata.ParamType.context:
                return context;
            case metadata.ParamType.context_request:
                return context.request;
            case metadata.ParamType.context_response:
                return context.response;
            case metadata.ParamType.context_next:
                return context.next;
            case metadata.ParamType.context_accept:
                return context.accept;
            case metadata.ParamType.context_accept_language:
                return context.language;
            default:
                throw Error('Invalid parameter type');
        }
    };
    InternalServer.prototype.convertType = function (paramValue, paramType) {
        var serializedType = paramType['name'];
        switch (serializedType) {
            case 'Number':
                return paramValue === undefined ? paramValue : parseFloat(paramValue);
            case 'Boolean':
                return paramValue === undefined ? paramValue : paramValue === 'true';
            default:
                return InternalServer.paramConverter(paramValue, paramType);
        }
    };
    InternalServer.serverClasses = new Map();
    InternalServer.paths = new Map();
    InternalServer.pathsResolved = false;
    InternalServer.serviceFactory = {
        create: function (serviceClass) {
            return new serviceClass();
        },
        getTargetClass: function (serviceClass) {
            return serviceClass;
        }
    };
    InternalServer.paramConverter = function (p, t) { return p; };
    return InternalServer;
}());
exports.InternalServer = InternalServer;
